### PR TITLE
Ref #952: Waitlists as relationships

### PR DESCRIPTION
**Pull request by leplatrem**
Ref #952

In this pull-request, we propose an implementation that leverages the new waitlists format in CTMS (see https://github.com/mozilla-it/ctms-api/pull/492)

The main idea is to be able to onboard waitlists with any arbitrary field, without having to redeploy Basket. 
The waitlists fields validation is fully managed by CTMS.

Basically, with this PR, when the following data is posted:
```
newsletters=vpn-bundle-waitlist
vpn_bundle_currency=eur
vpn_bundle_pack=gold
source_url=http://mozilla.org
```
The following payload is sent to CTMS:
```
"waitlists": [
  {
     "name": "vpn-bundle",
     "subscribed": True,
     "source": "http://mozilla.org",
     "fields": {"currency": "eur", "pack": "gold"},
  },
]
```

Onboarding a new waitlist would consist in:
1. Add new newsletter `abc-xyz-waitlist` with the `is_waitlist` flag set to True (name suffix is only by convention)
2. Create a form that sends `abc-xyz-waitlist` in the `newsletter` field, and optionally sends arbitrary data prefixed with `abc_xyz_`

### Documentation

* https://github.com/mozilla-it/ctms-api/blob/main/docs/adrs/waitlists_relationships.md
* https://mozilla-hub.atlassian.net/browse/CTMS-136
* old proposition about Basket inbox endpoint https://github.com/mozilla-it/ctms-api/pull/371

### Next steps

- [x] global approach is validated (see Rob comment)
- [x] unit tests pass
- [x] verify that vpn waitlist newsletter is called ~~`"fpn"`~~ `"guardian-vpn-waitlist"` in Basket database
- [x] verify that `newsletters` field is always sent in existing waitlist subscription forms (✅ relay waitlists, ✅ guardian vpn)
- [x] behaviour is validated using integration tests https://github.com/mozilla-it/ctms-api/pull/570
- [ ] this pull request is approved (review by commit should be easier)
- [x] CTMS new waitlist API is deployed and validated in production (it's retro compatible)
- [ ] this pull request is merged and deployed

### CODE REVIEW

**Review by robhudson — COMMENTED** [review]
Just a quick comment that the general approach here looks good to me. The model changes and migration, plus the edits based on the new boolean flag. 👍

**Comment by pmac** [comment]
I think we might have to change how the subscribe task works to make sure it doesn't drop the extra field data. I think it checks the incoming fields against a list of known fields. I could be wrong (I haven't looked) but wanted to call it out so we can check tomorrow.

**Comment by leplatrem** [comment]
Thanks for your early feedback, that's encouraging! 

I marked the pull-request as ready to review since I believe that all pieces are in place. 

There is no rush to merge/deploy, since the CTMS part is not yet deployed in production anyway.

File: basket/news/backends/ctms.py [hunk 1]
```diff
@@ -196,6 +206,49 @@ def truncate_string(max_length, raw_string):
...
+
+    is turned into::
+
+        {
+          "country": "fr",
+          "currency": "eur",
+        }
+    """
+    prefix = re.sub("[^0-9a-zA-Z]+", "_", slug) + "_"
+    input_data = data.copy()
```
Comment (leplatrem) [inline basket/news/backends/ctms.py:226]: TODO: this copy is not necessary

File: basket/news/backends/ctms.py [hunk 2]
```diff
@@ -103,11 +102,22 @@ def from_vendor(contact):
...
-            data["newsletters"] = newsletters
+                    data.setdefault("newsletters", []).append(newsletter["name"])
+        elif group_name == "waitlists":
+            # Unimported per-waitlist data: source, extra fields...
+            for waitlist in group:
+                wl_name = waitlist["name"]
+                # Legacy waitlist format. For backward compatibility.
+                # See `to_vendor()`` and `waitlist_fields_for_slug()` for the inverse.
+                if wl_name == "vpn":
+                    wl_name = "fpn"
```
Comment (leplatrem) [inline basket/news/backends/ctms.py:115]: slug is actually `guardian-vpn-waitlist`

**Comment by robhudson** [comment]
I can review this this week if it is all ready.

**Comment by leplatrem** [comment]
The code is ready to review yes. Thank you!
(not yet deployed in prod though)

**Comment by leplatrem** [comment]
@robhudson any chance you could have a look please? 🙏

File: basket/news/backends/ctms.py [hunk 3]
```diff
@@ -196,6 +208,49 @@ def truncate_string(max_length, raw_string):
...
+            consumed_key = name
+            name = new_fields[name]
+        else:
+            consumed_key = name
+        if not name.startswith(prefix):
+            continue
+        field_name = name.replace(prefix, "")
+        fields[field_name] = raw_value
+        consumed_keys.append(consumed_key)
+    print(slug, data, consumed_keys)
```
Comment (robhudson) [inline basket/news/backends/ctms.py:250]: We should remove `print` statements.

File: basket/news/backends/ctms.py [hunk 4]
```diff
@@ -103,11 +104,22 @@ def from_vendor(contact):
...
+                wl_name = waitlist["name"]
+                # Legacy waitlist format. For backward compatibility.
+                # See `to_vendor()`` and `waitlist_fields_for_slug()` for the inverse.
+                if wl_name == "vpn":
+                    wl_name = VPN_NEWSLETTER_SLUG
+                    data["fpn_country"] = waitlist["fields"]["geo"]
+                    data["fpn_platform"] = waitlist["fields"]["platform"]
+                if wl_name == "relay":
+                    data["relay_country"] = waitlist["fields"]["geo"]
+                data.setdefault("newsletters", []).append(wl_name)
```
Comment (robhudson) [inline basket/news/backends/ctms.py:122]: It's hard to tell from the diff here, but do you have tests for these `from_vendor` changes?

**Review by robhudson — CHANGES_REQUESTED** [review]
Just a couple things noted, otherwise this looks good to me.

File: basket/news/backends/ctms.py (hunk 4 above)
Comment (leplatrem) [reply basket/news/backends/ctms.py:124]: There's only one test for `from_vendor`, and yes it covers the changes.
https://github.com/mozmeao/basket/blob/2a3b77d7d2fe919d8e54d8e90d84e4e4a4202584/basket/news/tests/test_ctms.py#L130-L133

But thanks for your comment, it made discover an issue about newsletters names, that I fixed in the latests commits

**Comment by pmac** [comment]
Got a legit looking test failure, but other than that this is looking great. Thanks again!

**Comment by robhudson** [comment]
@leplatrem Any luck tracking down the test failure?

**Comment by leplatrem** [comment]
I am working on integration tests in CTMS that run basket. This will give us better confidence in this branch. I'll make sure the tests pass too.

**Comment by leplatrem** [comment]
I could confirm that integration tests pass. This would be ready to review :)

**Comment by leplatrem** [comment]
@robhudson up :)

File: basket/news/migrations/0026_newsletter_is_waitlist.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+def set_waitlist_field(apps, schema_editor):
+    """
+    Instead of hard-coding which newsletters are waitlists,
+    we now set a flag.
+    """
+    Newsletter = apps.get_model("news", "Newsletter")
+    for newsletter in Newsletter.objects.all():
+        is_waitlist = newsletter.slug.endswith("-waitlist") or newsletter.slug in {
+            "relay",
+            "fpn",
```
Comment (robhudson) [inline basket/news/migrations/0026_newsletter_is_waitlist.py:15]: Should this be `"vpn"`? Based on the above conditionals we check for the `"vpn"` newsletter and change it to `"guardian-vpn-waitlist"` but I've only seen `"fpn"` used as part of the extra data for the waitlist.

Comment (leplatrem) [reply basket/news/migrations/0026_newsletter_is_waitlist.py:15]: You are right, this conditional is superfluous. Looking at the list at https://basket.mozilla.org/news/ all waitlists slugs already end with `-waitlist`

**Comment by leplatrem** [comment]
CTMS v2 was deployed this morning. This can now be merge/deployed.

**Pull request merged**